server/src/main/java/org/opensearch/action/ActionModule.java

```diff
@@ -1112,41 +1115,49 @@ public boolean isActionRegistered(String actionName) {
         }
 
         /**
-         * Add a dynamic action to the registry.
+         * Adds a dynamic route to the registry.
          *
          * @param route The route instance to add
          * @param action The corresponding instance of RestSendToExtensionAction to execute
          */
-        public void registerDynamicRoute(RestHandler.Route route, RestSendToExtensionAction action) {
+        public void registerDynamicRoute(NamedRoute route, RestSendToExtensionAction action) {
             requireNonNull(route, "route is required");
             requireNonNull(action, "action is required");
-            Optional<String> routeName = Optional.empty();
-            if (route instanceof NamedRoute) {
-                routeName = Optional.of(((NamedRoute) route).name());
-                if (isActionRegistered(routeName.get()) || registeredActionNames.contains(routeName.get())) {
-                    throw new IllegalArgumentException("route [" + route + "] already registered");
-                }
+
+            Optional<String> routeName = Optional.of(route.name());
+            if (isActionRegistered(routeName.get())) {
+                throw new IllegalArgumentException("route [" + route + "] already registered");
             }
+            Set<String> actionNames = new HashSet<>(route.actionNames());
+
+            actionNames.forEach(act -> {
+                if (isActionRegistered(act)) {
```

Comment:
The idea is to check if any actionNames were already registered. In the approach you mentioned we are creating a duplicate set of registered entries to just check if there is any intersection, which creates new sets for every registration vs `isActionRegistered` which does  check in O(1) time. I prefer keeping it as is.

Reply:
Fair point, but my `Collections.disjoint()` suggestion is [the same](https://github.com/openjdk/jdk/blob/6ebb0e3bd4ba3579c66cdc5a329e95df7bda5b95/src/java.base/share/classes/java/util/Collections.java#L5786-L5836) iterate-and-check-contains solution.  The only problem is that it just returns a boolean rather than the matching action(s) so you still have to iterate to find the match to post in a log message; but you only get the first one.

Creating the new set helps give a better log message with all the conflicts, so you can defer that set intersection calculation until you need it.
```java
if (!Collections.disjoint(actionNames, registeredActionNames)) {
    Set<String> alreadyRegistered = new HashSet<>(registeredActionNames).retainAll(actionNames);
    String acts = alreadyRegistered.stream().collect(Collectors.joining(", ");
    throw new IllegalArgumentException("action"
        + (alreadyRegistered.size() > 1 ? "s [" : " [")
        + intersection 
        + "] already registered");
}
```

Reply:
done.